src/leap/mail/imap/account.py

```diff
@@ -248,66 +231,92 @@ def select(self, name, readwrite=1):
         :param readwrite: 1 for readwrite permissions.
         :type readwrite: int
 
-        :rtype: SoledadMailbox
+        :rtype: IMAPMailbox
         """
-        if PROFILE_CMD:
-            start = time.time()
+        name = normalize_mailbox(name)
 
-        name = self._parse_mailbox_name(name)
-        if name not in self.mailboxes:
-            logger.warning("No such mailbox!")
-            return None
-        self.selected = name
+        def check_it_exists(mailboxes):
+            if name not in mailboxes:
+                logger.warning("SELECT: No such mailbox!")
+                return None
+            return name
+
+        def set_selected(_):
+            self.selected = name
+
+        def get_collection(name):
+            if name is None:
+                return None
+            return self.account.get_collection_by_mailbox(name)
 
-        sm = SoledadMailbox(
-            name, self._soledad, self._memstore, readwrite)
-        if PROFILE_CMD:
-            _debugProfiling(None, "SELECT", start)
-        return sm
+        d = self.account.list_all_mailbox_names()
+        d.addCallback(check_it_exists)
+        d.addCallback(get_collection)
+        d.addCallback(partial(
+            self._return_mailbox_from_collection, readwrite=readwrite))
+        return d
 
     def delete(self, name, force=False):
         """
         Deletes a mailbox.
 
-        Right now it does not purge the messages, but just removes the mailbox
-        name from the mailboxes list!!!
-
         :param name: the mailbox to be deleted
         :type name: str
 
-        :param force: if True, it will not check for noselect flag or inferior
-                      names. use with care.
+        :param force:
+            if True, it will not check for noselect flag or inferior
+            names. use with care.
         :type force: bool
+        :rtype: Deferred
         """
-        name = self._parse_mailbox_name(name)
+        name = normalize_mailbox(name)
+        _mboxes = None
+
+        def check_it_exists(mailboxes):
+            global _mboxes
+            _mboxes = mailboxes
+            if name not in mailboxes:
+                err = imap4.MailboxException("No such mailbox: %r" % name)
+                return defer.fail(err)
```

Comment:
changed, this suffered from progressive conversion to deferreds, but now it's ok